pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_index.go

```diff
@@ -0,0 +1,438 @@
+				if start != 0 {
+					panic(fmt.Sprintf("Index (%v) has more than one index partitioning. Should be exactly one", ie.IndexID))
+				}
+				start = int(t.NumImplicitColumns)
+			}
+		}
+	})
+	// Currently, we only allow implicit partitioning on hash sharded index. When
+	// that happens, the shard column always comes after implicit partition
+	// columns.
+	if ie.Sharding != nil && ie.Sharding.IsSharded {
+		start++
+	}
+	return start
+}
+
+// explicitColumnIDsWithoutShardColumn retrieve explicit column ID (excluding shard column)
+// of index element `e`.
+func explicitColumnIDsWithoutShardColumn(b BuildCtx, ie *scpb.Index) descpb.ColumnIDs {
+	explicitColIDs := ie.KeyColumnIDs[explicitColumnStartIdx(b, ie):]
+	explicitColNames := make([]string, len(explicitColIDs))
+	for i, colID := range explicitColIDs {
+		colElms := b.ResolveColumnByID(ie.TableID, colID, ResolveParams{
+			IsExistenceOptional: false,
+			RequiredPrivilege:   privilege.CREATE,
+		})
+		_, _, cne := scpb.FindColumnName(colElms)
+		if cne == nil {
+			panic(fmt.Sprintf("No column name is found for column ID %v", colID))
+		}
+		explicitColNames[i] = cne.Name
+	}
+
+	colIDs := make(descpb.ColumnIDs, 0, len(explicitColIDs))
+	for i := range explicitColNames {
+		if ie.Sharding == nil || !ie.Sharding.IsSharded || explicitColNames[i] != ie.Sharding.Name {
+			colIDs = append(colIDs, explicitColIDs[i])
+		}
+	}
+	return colIDs
+}
+
+// isValidReferencedUniqueConstraint return true if `e` is a unique index and this index's
+// explicit columns are a permutation of `referencedColIDs`.
+func isValidReferencedUniqueConstraint(
+	b BuildCtx, ie *scpb.Index, referencedColIDs []tree.ColumnID,
+) bool {
+	return ie.IsUnique && !ie.IsPartial &&
+		explicitColumnIDsWithoutShardColumn(b, ie).PermutationOf(referencedColIDs)
+}
+
+// uniqueConstraintHasReplacementCandidate returns true if `elms` contains an index
+// that can serve as a replacement candidate for the to-be-dropped secondary index `sie`, which
+// references columns `referencedColumnIDs`.
+func uniqueConstraintHasReplacementCandidate(
+	b BuildCtx,
+	elms ElementResultSet,
+	sie *scpb.SecondaryIndex,
+	referencedColumnIDs []descpb.ColumnID,
+) bool {
+	result := false
+
+	// Check all indexes (both primary and secondary) to see if we can find a replacement candidate.
+	elms.ForEachElementStatus(func(current scpb.Status, target scpb.TargetStatus, e scpb.Element) {
+		if result {
+			return
+		}
+
+		switch t := e.(type) {
+		case *scpb.SecondaryIndex:
+			if t.IndexID != sie.IndexID && isValidReferencedUniqueConstraint(b, &t.Index, referencedColumnIDs) {
+				result = true
+			}
+		case *scpb.PrimaryIndex:
+			if isValidReferencedUniqueConstraint(b, &t.Index, referencedColumnIDs) {
+				result = true
+			}
+		}
+	})
+
+	return result
+}
+
+// Attempt to drop all views that depend on the to be dropped index if CASCADE.
+// Panic if there is a dependent view but drop behavior is not CASCADE.
+func mayDropDependentViews(
```

Comment:
personally I find `maybe` to sound better than `may` when writing logic which is going to check some condition and then either fail or do a thing.
----
Aside: I do like this pattern, for what it's worth. When I first started programming, it smelled to me. I would think to myself, shouldn't the condition checking happen above calling the function which does a thing? Having conditional logic in multi-step functions however, can muddle the readability, especially when the condition checking is cumbersome.